lens-cube/src/main/java/org/apache/lens/cube/parse/MaxCoveringFactResolver.java

```diff
@@ -31,193 +30,174 @@
+          }
+        }
+      }
+    }
+    //  cubeql.pruneCandidateFactWithCandidateSet(CandidateTablePruneCause.lessData(null));
 
   }
-  public MaxCoveringFactResolver(Configuration conf) {
+
+  private void resolveByDataCompleteness(CubeQueryContext cubeql) {
+    // From the list of  candidate fact sets, we calculate the maxDataCompletenessFactor.
+    float maxDataCompletenessFactor = 0f;
+    for (Candidate cand : cubeql.getCandidates()) {
+      float dataCompletenessFactor = computeDataCompletenessFactor(cand);
+      if (dataCompletenessFactor > maxDataCompletenessFactor) {
+        maxDataCompletenessFactor = dataCompletenessFactor;
+      }
+    }
+
+    if (maxDataCompletenessFactor == 0f) {
+      //there is nothing to prune
+      return;
+    }
+
+    // We prune those candidate fact set, whose dataCompletenessFactor is less than maxDataCompletenessFactor
+    Iterator<Candidate> iter = cubeql.getCandidates().iterator();
+    while (iter.hasNext()) {
+      Candidate cand = iter.next();
+      float dataCompletenessFactor = computeDataCompletenessFactor(cand);
+      if (dataCompletenessFactor < maxDataCompletenessFactor) {
+        log.info("Not considering Candidate :{} from the list as the dataCompletenessFactor for this:{} is "
+          + "less than the max:{}", cand, dataCompletenessFactor, maxDataCompletenessFactor);
+        iter.remove();
+        cubeql.addCandidatePruningMsg(cand,
+          new CandidateTablePruneCause(CandidateTablePruneCause.CandidateTablePruneCode.INCOMPLETE_PARTITION));
+      }
+    }
+  }
+
+  private float computeDataCompletenessFactor(Candidate cand) {
+    float completenessFactor = 0f;
+    int numPartition = 0;
+    for (StorageCandidate sc : CandidateUtil.getStorageCandidates(cand)) {
+      if (sc.getDataCompletenessMap() != null) {
+        Map<String, Map<String, Float>> completenessMap = sc.getDataCompletenessMap();
+        for (Map<String, Float> partitionCompleteness : completenessMap.values()) {
+          for (Float value : partitionCompleteness.values()) {
+            numPartition++;
+            completenessFactor += value;
+          }
+        }
+      }
+    }
+    return numPartition == 0 ? completenessFactor : completenessFactor / numPartition;
   }
 
+  /**
+   * Returns time covered by fact set for each part column.
+   *
+   * @param cand
+   * @return
+   */
+  private Map<String, Long> getTimeCoveredForEachPartCol(Candidate cand) {
+    Map<String, Long> ret = Maps.newHashMap();
+    UpdatePeriod smallest = UpdatePeriod.values()[UpdatePeriod.values().length - 1];
+    for (FactPartition part : cand.getParticipatingPartitions()) {
+      if (part.getPeriod().compareTo(smallest) < 0) {
+        smallest = part.getPeriod();
+      }
+    }
+    PartitionRangesForPartitionColumns partitionRangesForPartitionColumns = new PartitionRangesForPartitionColumns();
+    for (FactPartition part : cand.getParticipatingPartitions()) {
```

Comment:
can we do cand.getParticipatingPartitions() operation only once . Each call will have to traverse the candidate tree and get partitions.